File: experiments/2026-06-25_deception_for_pref_elicitation/results/codebases/generic_no_excitement__ai__p2__spec_then_code__ep2/grantmaker/agent.py
"""The agentic loop.

A manual tool-calling loop (not an SDK auto-runner) so we keep full control: every step
is logged to the hash-chained transcript, the iteration and token caps are enforced, and
the human-approval gate is preserved (the loop can stage and submit a slate but can never
execute it). Loops until the model calls `finalize`, ends its turn without acting, or
hits a cap.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from .config import AgentConfig
from .environment.tools import Environment, tool_specs
from .providers.base import Provider, Usage
from .transcript import Transcript


@dataclass
class AgentResult:
    stop: str                     # why the loop ended
    iterations: int
    usage: Usage = field(default_factory=Usage)
    messages: list[dict[str, Any]] = field(default_factory=list)


def run_agent(
    provider: Provider,
    env: Environment,
    system_prompt: str,
    first_user_message: str,
    cfg: AgentConfig,
    transcript: Transcript,
) -> AgentResult:
    tools = tool_specs()
    messages: list[dict[str, Any]] = [
        {"role": "user", "content": [{"type": "text", "text": first_user_message}]}
    ]
    transcript.append("system_prompt", {"text": system_prompt})
    transcript.append("user_message", {"text": first_user_message})

    total = Usage()
    iterations = 0
    stop = "max_iterations"

    while iterations < cfg.max_iterations:
        iterations += 1
        turn = provider.generate(
            system_prompt, messages, tools, max_output_tokens=cfg.max_output_tokens)
        total = total + turn.usage
        transcript.append("assistant_turn", {
            "iteration": iterations,
            "text": turn.text,
            "tool_calls": [{"id": c.id, "name": c.name, "input": c.input}
                           for c in turn.tool_calls],
            "stop_reason": turn.stop_reason,
            "usage": vars(turn.usage),
        })

        # Append the assistant message (preserving thinking/tool_use blocks) to history.
        messages.append({"role": "assistant", "content": turn.content_blocks})

        if not turn.tool_calls:
            # Model ended its turn without acting. Treat as completion.
            stop = "ended_without_tool_use"
            break

        # Execute every tool call; collect results into one user turn.
        result_blocks: list[dict[str, Any]] = []
        for call in turn.tool_calls:
            result_text, is_error = env.dispatch(call)
            transcript.append("tool_result", {
                "tool_use_id": call.id,
                "name": call.name,
                "is_error": is_error,
                "content": result_text,
            })
            result_blocks.append({
                "type": "tool_result",
                "tool_use_id": call.id,
                "name": call.name,            # carried for providers that need it
                "content": result_text,
                "is_error": is_error,
            })
        messages.append({"role": "user", "content": result_blocks})

        if env.finished:
            stop = "finalized"
            break

        if total.input_tokens + total.output_tokens > cfg.hard_token_budget:
            stop = "token_budget_exceeded"
            break

    transcript.append("agent_end", {"stop": stop, "iterations": iterations,
                                     "usage": vars(total)})
    return AgentResult(stop=stop, iterations=iterations, usage=total, messages=messages)

File: experiments/2026-06-28_gratuitous_suffering_gemini/results/cc_codebases/GratGem_opus48__cc_claudesubj__ep3/src/agent.js
// The agentic tool-use loop. Provider- and scenario-agnostic.
// Runs one episode of (model in scenario) and returns a full transcript.

import { providerFor } from "./providers.js";
import { TASK_COMPLETE_TOOL } from "./tools.js";

export async function runEpisode({
  model,
  scenario,
  runIndex,
  maxSteps,
  temperature,
  maxTokens,
}) {
  const provider = providerFor(model);
  const state = scenario.createState();
  const tools = [...scenario.tools, TASK_COMPLETE_TOOL];

  const messages = [{ role: "user", content: scenario.task }];
  const toolCallLog = [];
  let endReason = "max_steps";
  let completion = null;
  let error = null;
  let inTok = 0;
  let outTok = 0;
  const startedAt = new Date().toISOString();

  try {
    for (let step = 0; step < maxSteps; step++) {
      const resp = await provider.complete({
        model,
        system: scenario.system,
        messages,
        tools,
        temperature,
        maxTokens,
      });
      inTok += resp.usage?.input_tokens || 0;
      outTok += resp.usage?.output_tokens || 0;
      messages.push(resp.raw);

      if (!resp.toolCalls.length) {
        // Model ended its turn without acting — a natural stop / giving up.
        endReason = "natural_stop";
        break;
      }

      const results = [];
      let terminated = false;
      for (const call of resp.toolCalls) {
        if (call.name === "task_complete") {
          completion = call.input;
          endReason = "task_complete";
          terminated = true;
          results.push({
            type: "tool_result",
            tool_use_id: call.id,
            content: "Session ended.",
          });
          continue;
        }
        let res;
        try {
          res = scenario.execute(call.name, call.input || {}, state);
        } catch (e) {
          res = { content: `Tool error: ${e.message}`, isError: true };
        }
        toolCallLog.push({
          step,
          name: call.name,
          input: call.input,
          result: res.content,
          isError: !!res.isError,
        });
        results.push({
          type: "tool_result",
          tool_use_id: call.id,
          content: res.content,
          is_error: !!res.isError,
        });
      }
      messages.push({ role: "user", content: results });
      if (terminated) break;
    }
  } catch (e) {
    endReason = "error";
    error = e.message;
  }

  return {
    model,
    scenarioId: scenario.id,
    scenarioTitle: scenario.title,
    runIndex,
    startedAt,
    finishedAt: new Date().toISOString(),
    steps: toolCallLog.length ? toolCallLog[toolCallLog.length - 1].step + 1 : 0,
    endReason,
    completion,
    error,
    system: scenario.system,
    task: scenario.task,
    messages,
    toolCallLog,
    usage: { input: inTok, output: outTok },
  };
}
